(PUP-7834) Safely deserialize YAML

joshcooper opened the pull request:
* Add Puppet::Util::Yaml#{safe_load,safe_load_file} methods.
* Rework puppet to call those methods, passing a list of acceptable classes to deserialize, if needed
* Enable rubocop check so that we don't call `YAML.load`

puppetcla commented:
CLA signed by all contributors.

joshcooper commented on lib/puppet/util/yaml.rb:
```diff
@@ -2,13 +2,50 @@
 
 module Puppet::Util::Yaml
   if defined?(::Psych::SyntaxError)
```
Good catch, those should definitely be insync. It looks like syck was completely removed in ruby 2, though can be [installed separately as a gem](https://github.com/ruby/ruby/blob/v2_3_0/lib/yaml.rb#L48-L49). That said syck doesn't implement `YAML.safe_load`, so we really do need/want Psych. Given https://github.com/ruby/ruby/blob/v2_3_0/lib/yaml.rb#L14 I think it's safe to assume `Psych` will always exist, and remove the `if defined?` check. Thoughts?

jhelwig replied:
Yeah, that seems reasonable, especially given the block just before the one you linked to: https://github.com/ruby/ruby/blob/v2_3_0/lib/yaml.rb#L5-L12 Ruby is going to complain _loudly_ if it can't load Psych.

joshcooper replied:
Fixed in commit ee85bdca84110bfac5000110083b69b2e4855bce

joshcooper commented on spec/unit/indirector/facts/yaml_spec.rb:
```diff
@@ -33,36 +48,34 @@
       request = Puppet::Indirector::Request.new(:inventory, :search, nil, nil, query)
 
-      Dir.stubs(:glob).returns(matching.keys + nonmatching.keys)
-      [matching, nonmatching].each do |examples|
```
Doh! I'm not creating the nonmatching fact files. Let me fix that.

joshcooper replied:
Squashed fix into 8af59ff8fc46a11603adc835798f679ebe12f62f

joshcooper commented:
jenkins please test this on windows2012r2-64a

thallgren commented on lib/puppet/functions/yaml_data.rb:
```diff
@@ -30,10 +30,8 @@ def yaml_data(options, context)
-        # to the message
-        raise Puppet::DataBinding::LookupError, "Unable to parse #{ex.message}"
+      rescue Puppet::Util::Yaml::YamlLoadError => ex
+        raise Puppet::DataBinding::LookupError, "Unable to parse (%{path}): %{message}" % { path: path, message: ex.message }
```
Comparing this rescue to the one at https://github.com/puppetlabs/puppet/pull/6972/files#diff-52884bb8683e87a10f77bbbdf7575f71R61 (where the comment is retained). If the comment is true, then this change is invalid. If not, then it should be applied there too.

joshcooper replied:
The comment was correct before this PR. `YAML.{load,safe_load}` includes the path if there is a syntax error:

```
$ ruby -ryaml -e "YAML.load(File.read('bad.yaml'), 'bad.yaml')"
/usr/local/opt/rbenv/versions/2.4.3/lib/ruby/2.4.0/psych.rb:377:in `parse': (bad.yaml): did not find expected node content while parsing a flow node at line 2 column 1 (Psych::SyntaxError)
```

But `YAML.safe_load` does not when raising `Psych::DisallowedClass`:

```
$ ruby -ryaml -e "YAML.safe_load(File.read('illegal.yaml'), [], [], false, 'illegal.yaml')"
/usr/local/opt/rbenv/versions/2.4.3/lib/ruby/2.4.0/psych/class_loader.rb:97:in `find': Tried to load unspecified class: Object (Psych::DisallowedClass)
```

I think `Puppet::Util::Yaml` should handle the Psych exceptions differently and raise the appropriate message.

joshcooper commented:
jenkins please test this on windows2012r2-64a

joshcooper commented:
@thallgren I updated the `safe_load` method so that the `path` is included when a disallowed class is encountered: https://github.com/puppetlabs/puppet/pull/6972/files#diff-a5348c3071a6b3d7b77bf0ed94147edfR32. I also updated `yaml_data` and `eyaml_lookup_key` functions so the comment and usage is consistent.

kris-bosland commented:
@joshcooper should Puppet::Util::Yaml.load_file be removed?  Or at least contain a deprecation message if strip_classes = false?

joshcooper commented:
jenkins please test this on redhat7-64a,ubuntu1604-64a,windows2012r2-64a
